File: lib/imagekit_ffi.dart
import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';

// ignore: depend_on_referenced_packages
import 'package:ffi/ffi.dart';

import 'imagekit_ffi_bindings_generated.dart';

const String _libName = 'imagekit_ffi';

final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

final ImagekitFfiBindings _bindings = ImagekitFfiBindings(_dylib);

class ImageKitFfi {
  /// Освобождение выделенного нативного буфера
  void freeBuffer(Pointer<Uint8> buffer) {
    if (buffer != nullptr) {
      _bindings.free_buffer(buffer);
    }
  }

  Uint8List convertYuv420ToJpegBuffer({
    required Uint8List yPlane,
    required Uint8List uPlane,
    required Uint8List vPlane,
    required int width,
    required int height,
    required int yStride,
    required int uStride,
    required int vStride,
    int uPixStride = 1,
    int vPixStride = 1,
    int rotationDegrees = 0,
    int quality = 90,
  }) {
    final yPtr = malloc<Uint8>(yPlane.length);
    final uPtr = malloc<Uint8>(uPlane.length);
    final vPtr = malloc<Uint8>(vPlane.length);
    final outSizePtr = malloc<Size>();
    Pointer<Uint8> jpegPtr = nullptr;

    try {
      yPtr.asTypedList(yPlane.length).setAll(0, yPlane);
      uPtr.asTypedList(uPlane.length).setAll(0, uPlane);
      vPtr.asTypedList(vPlane.length).setAll(0, vPlane);

      jpegPtr = _bindings.convert_yuv420_to_jpeg(
        yPtr,
        uPtr,
        vPtr,
        width,
        height,
        yStride,
        uStride,
        vStride,
        uPixStride,
        vPixStride,
        rotationDegrees,
        quality,
        outSizePtr,
      );

      final jpegSize = outSizePtr.value;
      if (jpegPtr == nullptr || jpegSize <= 0) {
        throw Exception('YUV420 -> JPEG conversion failed');
      }

      return Uint8List.fromList(jpegPtr.asTypedList(jpegSize));
    } finally {
      malloc.free(yPtr);
      malloc.free(uPtr);
      malloc.free(vPtr);
      malloc.free(outSizePtr);
      if (jpegPtr != nullptr) {
        _bindings.free_buffer(jpegPtr);
      }
    }
  }

  Uint8List convertNv21ToJpegBuffer({
    required Uint8List yPlane,
    required Uint8List uvPlane,
    required int width,
    required int height,
    required int yStride,
    required int uvStride,
    int uvPixStride = 2,
    int rotationDegrees = 0,
    int quality = 90,
  }) {
    final yPtr = malloc<Uint8>(yPlane.length);
    final uvPtr = malloc<Uint8>(uvPlane.length);
    final outSizePtr = malloc<Size>();
    Pointer<Uint8> jpegPtr = nullptr;

    try {
      yPtr.asTypedList(yPlane.length).setAll(0, yPlane);
      uvPtr.asTypedList(uvPlane.length).setAll(0, uvPlane);

      jpegPtr = _bindings.convert_nv21_to_jpeg(
        yPtr,
        uvPtr,
        width,
        height,
        yStride,
        uvStride,
        uvPixStride,
        rotationDegrees,
        quality,
        outSizePtr,
      );

      final jpegSize = outSizePtr.value;
      if (jpegPtr == nullptr || jpegSize <= 0) {
        throw Exception('NV21 -> JPEG conversion failed');
      }

      return Uint8List.fromList(jpegPtr.asTypedList(jpegSize));
    } finally {
      malloc.free(yPtr);
      malloc.free(uvPtr);
      malloc.free(outSizePtr);
      if (jpegPtr != nullptr) {
        _bindings.free_buffer(jpegPtr);
      }
    }
  }

  Uint8List encodeRgbaToJpegBuffer({
    required Uint8List rgbaBytes,
    required int width,
    required int height,
    int quality = 90,
  }) {
    final rgbaPtr = malloc<Uint8>(rgbaBytes.length);
    final outSizePtr = malloc<Size>();
    Pointer<Uint8> jpegPtr = nullptr;

    try {
      rgbaPtr.asTypedList(rgbaBytes.length).setAll(0, rgbaBytes);

      jpegPtr = _bindings.encode_rgba_to_jpeg_buffer(
        rgbaPtr,
        width,
        height,
        quality,
        outSizePtr,
      );

      final jpegSize = outSizePtr.value;
      if (jpegPtr == nullptr || jpegSize <= 0) {
        throw Exception('RGBA -> JPEG conversion failed');
      }

      return Uint8List.fromList(jpegPtr.asTypedList(jpegSize));
    } finally {
      malloc.free(rgbaPtr);
      malloc.free(outSizePtr);
      if (jpegPtr != nullptr) {
        _bindings.free_buffer(jpegPtr);
      }
    }
  }

  /// Поворот RGBA изображения
  Uint8List rotateRgbaImage({
    required Uint8List rgbaBytes,
    required int width,
    required int height,
    required int rotationDegrees,
  }) {
    final rgbaPtr = malloc<Uint8>(rgbaBytes.length);
    Pointer<Uint8> rotatedPtr = nullptr;

    try {
      rgbaPtr.asTypedList(rgbaBytes.length).setAll(0, rgbaBytes);

      rotatedPtr = _bindings.rotate_rgba_image(
        rgbaPtr,
        width,
        height,
        rotationDegrees,
      );

      if (rotatedPtr == nullptr) {
        throw Exception('RGBA rotation failed');
      }

      final rotatedWidth = (rotationDegrees % 180 == 0) ? width : height;
      final rotatedHeight = (rotationDegrees % 180 == 0) ? height : width;
      final outputLength = rotatedWidth * rotatedHeight * 4;

      return Uint8List.fromList(rotatedPtr.asTypedList(outputLength));
    } finally {
      malloc.free(rgbaPtr);
      if (rotatedPtr != nullptr) {
        _bindings.free_buffer(rotatedPtr);
      }
    }
  }

  Uint8List encodeBgraToJpegBuffer({
    required Uint8List bgraBytes,
    required int width,
    required int height,
    int quality = 90,
  }) {
    final bgraPtr = malloc<Uint8>(bgraBytes.length);
    final outSizePtr = malloc<Size>();
    Pointer<Uint8> jpegPtr = nullptr;

    try {
      bgraPtr.asTypedList(bgraBytes.length).setAll(0, bgraBytes);

      jpegPtr = _bindings.encode_bgra_to_jpeg_buffer(
        bgraPtr,
        width,
        height,
        quality,
        outSizePtr,
      );

      final jpegSize = outSizePtr.value;
      if (jpegPtr == nullptr || jpegSize <= 0) {
        throw Exception('BGRA -> JPEG conversion failed');
      }

      return Uint8List.fromList(jpegPtr.asTypedList(jpegSize));
    } finally {
      malloc.free(bgraPtr);
      malloc.free(outSizePtr);
      if (jpegPtr != nullptr) {
        _bindings.free_buffer(jpegPtr);
      }
    }
  }

  Uint8List rotateJpegBuffer({
    required Uint8List jpegBytes,
    required int rotationDegrees,
  }) {
    final jpegPtr = malloc<Uint8>(jpegBytes.length);
    final outSizePtr = malloc<Size>();
    Pointer<Uint8> rotatedPtr = nullptr;

    try {
      jpegPtr.asTypedList(jpegBytes.length).setAll(0, jpegBytes);

      rotatedPtr = _bindings.rotate_jpeg_buffer(
        jpegPtr,
        jpegBytes.length,
        rotationDegrees,
        outSizePtr,
      );

      final outSize = outSizePtr.value;
      if (rotatedPtr == nullptr || outSize <= 0) {
        throw Exception('JPEG rotation failed');
      }

      return Uint8List.fromList(rotatedPtr.asTypedList(outSize));
    } finally {
      malloc.free(jpegPtr);
      malloc.free(outSizePtr);
      if (rotatedPtr != nullptr) {
        _bindings.free_buffer(rotatedPtr);
      }
    }
  }
}

File: src/imagekit_ffi.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <turbojpeg.h>
#include "imagekit_ffi.h"

#ifdef _WIN32
#define EXPORT __declspec(dllexport)
#else
#define EXPORT __attribute__((visibility("default")))
#endif

void YUV_to_RGB(int Y, int U, int V, uint8_t *R, uint8_t *G, uint8_t *B)
{
  int C = Y - 16;
  int D = U - 128;
  int E = V - 128;

  int r = (298 * C + 409 * E + 128) >> 8;
  int g = (298 * C - 100 * D - 208 * E + 128) >> 8;
  int b = (298 * C + 516 * D + 128) >> 8;

  *R = (uint8_t)(r < 0 ? 0 : (r > 255 ? 255 : r));
  *G = (uint8_t)(g < 0 ? 0 : (g > 255 ? 255 : g));
  *B = (uint8_t)(b < 0 ? 0 : (b > 255 ? 255 : b));
}

EXPORT void free_buffer(uint8_t *buffer)
{
  if (buffer)
    free(buffer);
}

EXPORT uint8_t *rotate_rgba_image(const uint8_t *rgba, int width, int height, int rotationDegrees)
{
  if (!rgba || width <= 0 || height <= 0 ||
      (rotationDegrees != 0 && rotationDegrees != 90 && rotationDegrees != 180 && rotationDegrees != 270))
  {
    return NULL;
  }

  int outWidth = (rotationDegrees == 90 || rotationDegrees == 270) ? height : width;
  int outHeight = (rotationDegrees == 90 || rotationDegrees == 270) ? width : height;

  uint8_t *rotated = (uint8_t *)malloc(outWidth * outHeight * 4);
  if (!rotated)
    return NULL;

  for (int y = 0; y < height; y++)
  {
    for (int x = 0; x < width; x++)
    {
      int src_idx = (y * width + x) * 4;
      int rotated_x, rotated_y;
      switch (rotationDegrees)
      {
      case 90:
        rotated_x = height - 1 - y;
        rotated_y = x;
        break;
      case 180:
        rotated_x = width - 1 - x;
        rotated_y = height - 1 - y;
        break;
      case 270:
        rotated_x = y;
        rotated_y = width - 1 - x;
        break;
      default:
        rotated_x = x;
        rotated_y = y;
        break;
      }
      int dst_idx = (rotated_y * outWidth + rotated_x) * 4;
      memcpy(&rotated[dst_idx], &rgba[src_idx], 4);
    }
  }

  return rotated;
}

EXPORT uint8_t *convert_yuv420_to_rgba(const uint8_t *y_plane, const uint8_t *u_plane, const uint8_t *v_plane,
                                       int width, int height,
                                       int y_stride, int u_stride, int v_stride,
                                       int u_pix_stride, int v_pix_stride,
                                       int rotationDegrees)
{
  uint8_t *intermediate = (uint8_t *)malloc(width * height * 4);
  if (!intermediate)
    return NULL;

  for (int y = 0; y < height; ++y)
  {
    for (int x = 0; x < width; ++x)
    {
      int y_val = y_plane[y * y_stride + x];
      int u_val = u_plane[(y / 2) * u_stride + (x / 2) * u_pix_stride];
      int v_val = v_plane[(y / 2) * v_stride + (x / 2) * v_pix_stride];

      uint8_t R, G, B;
      YUV_to_RGB(y_val, u_val, v_val, &R, &G, &B);

      int rgba_idx = (y * width + x) * 4;
      intermediate[rgba_idx + 0] = R;
      intermediate[rgba_idx + 1] = G;
      intermediate[rgba_idx + 2] = B;
      intermediate[rgba_idx + 3] = 255;
    }
  }

  uint8_t *rotated = rotate_rgba_image(intermediate, width, height, rotationDegrees);
  free(intermediate);
  return rotated;
}

EXPORT uint8_t *convert_nv21_to_rgba(const uint8_t *y_plane, const uint8_t *uv_plane,
                                     int width, int height,
                                     int y_stride, int uv_stride, int uv_pix_stride,
                                     int rotationDegrees)
{
  uint8_t *intermediate = (uint8_t *)malloc(width * height * 4);
  if (!intermediate)
    return NULL;

  for (int y = 0; y < height; ++y)
  {
    for (int x = 0; x < width; ++x)
    {
      int y_val = y_plane[y * y_stride + x];
      int uv_x = (x / 2) * uv_pix_stride;
      int uv_y = (y / 2) * uv_stride;
      int v_val = uv_plane[uv_y + uv_x];
      int u_val = uv_plane[uv_y + uv_x + 1];

      uint8_t R, G, B;
      YUV_to_RGB(y_val, u_val, v_val, &R, &G, &B);

      int rgba_idx = (y * width + x) * 4;
      intermediate[rgba_idx + 0] = R;
      intermediate[rgba_idx + 1] = G;
      intermediate[rgba_idx + 2] = B;
      intermediate[rgba_idx + 3] = 255;
    }
  }

  uint8_t *rotated = rotate_rgba_image(intermediate, width, height, rotationDegrees);
  free(intermediate);
  return rotated;
}

EXPORT uint8_t *encode_rgba_to_jpeg_buffer(const uint8_t *rgba, int width, int height, int quality, size_t *out_size)
{
  tjhandle handle = tjInitCompress();
  if (!handle)
    return NULL;

  uint8_t *rgb = (uint8_t *)malloc(width * height * 3);
  if (!rgb)
  {
    tjDestroy(handle);
    return NULL;
  }

  for (int i = 0, j = 0; i < width * height * 4; i += 4, j += 3)
  {
    rgb[j] = rgba[i];
    rgb[j + 1] = rgba[i + 1];
    rgb[j + 2] = rgba[i + 2];
  }

  unsigned char *jpegBuf = NULL;
  unsigned long jpegSize = 0;

  int success = tjCompress2(handle, rgb, width, 0, height, TJPF_RGB,
                            &jpegBuf, &jpegSize, TJSAMP_444, quality, TJFLAG_ACCURATEDCT);

  free(rgb);
  tjDestroy(handle);

  if (success != 0)
  {
    if (jpegBuf)
      tjFree(jpegBuf);
    return NULL;
  }

  *out_size = jpegSize;
  return jpegBuf;
}

EXPORT uint8_t *convert_yuv420_to_jpeg(const uint8_t *y_plane, const uint8_t *u_plane, const uint8_t *v_plane,
                                       int width, int height,
                                       int y_stride, int u_stride, int v_stride,
                                       int u_pix_stride, int v_pix_stride,
                                       int rotationDegrees,
                                       int quality,
                                       size_t *out_size)
{
  uint8_t *rgba = convert_yuv420_to_rgba(y_plane, u_plane, v_plane,
                                         width, height,
                                         y_stride, u_stride, v_stride,
                                         u_pix_stride, v_pix_stride,
                                         rotationDegrees);
  if (!rgba)
    return NULL;

  int out_width = (rotationDegrees == 90 || rotationDegrees == 270) ? height : width;
  int out_height = (rotationDegrees == 90 || rotationDegrees == 270) ? width : height;

  uint8_t *jpeg = encode_rgba_to_jpeg_buffer(rgba, out_width, out_height, quality, out_size);
  free(rgba);
  return jpeg;
}

EXPORT uint8_t *convert_nv21_to_jpeg(const uint8_t *y_plane, const uint8_t *uv_plane,
                                     int width, int height,
                                     int y_stride, int uv_stride, int uv_pix_stride,
                                     int rotationDegrees,
                                     int quality,
                                     size_t *out_size)
{
  uint8_t *rgba = convert_nv21_to_rgba(y_plane, uv_plane,
                                       width, height,
                                       y_stride, uv_stride, uv_pix_stride,
                                       rotationDegrees);
  if (!rgba)
    return NULL;

  int out_width = (rotationDegrees == 90 || rotationDegrees == 270) ? height : width;
  int out_height = (rotationDegrees == 90 || rotationDegrees == 270) ? width : height;

  uint8_t *jpeg = encode_rgba_to_jpeg_buffer(rgba, out_width, out_height, quality, out_size);
  free(rgba);
  return jpeg;
}

EXPORT uint8_t *encode_bgra_to_jpeg_buffer(const uint8_t *bgra, int width, int height, int quality, size_t *out_size)
{
  if (!bgra || width <= 0 || height <= 0 || !out_size)
    return NULL;

  tjhandle handle = tjInitCompress();
  if (!handle)
    return NULL;

  uint8_t *rgb = (uint8_t *)malloc(width * height * 3);
  if (!rgb)
  {
    tjDestroy(handle);
    return NULL;
  }

  for (int i = 0, j = 0; i < width * height * 4; i += 4, j += 3)
  {
    rgb[j] = bgra[i + 2];
    rgb[j + 1] = bgra[i + 1];
    rgb[j + 2] = bgra[i + 0];
  }

  unsigned char *jpegBuf = NULL;
  unsigned long jpegSize = 0;

  int success = tjCompress2(handle, rgb, width, 0, height, TJPF_RGB,
                            &jpegBuf, &jpegSize, TJSAMP_444, quality, TJFLAG_ACCURATEDCT);

  free(rgb);
  tjDestroy(handle);

  if (success != 0)
  {
    if (jpegBuf)
      tjFree(jpegBuf);
    return NULL;
  }

  *out_size = jpegSize;
  return jpegBuf;
}

EXPORT uint8_t *rotate_jpeg_buffer(const uint8_t *jpeg_data, size_t jpeg_size, int rotationDegrees, size_t *out_size)
{
  if (!jpeg_data || jpeg_size == 0 || !out_size ||
      (rotationDegrees != 0 && rotationDegrees != 90 && rotationDegrees != 180 && rotationDegrees != 270))
  {
    return NULL;
  }

  tjhandle decompressHandle = tjInitDecompress();
  int width, height, jpegSubsamp, jpegColorspace;
  if (tjDecompressHeader3(decompressHandle, jpeg_data, jpeg_size, &width, &height, &jpegSubsamp, &jpegColorspace) != 0)
  {
    tjDestroy(decompressHandle);
    return NULL;
  }

  unsigned char *rgbaBuf = (unsigned char *)malloc(width * height * 4);
  if (!rgbaBuf)
  {
    tjDestroy(decompressHandle);
    return NULL;
  }

  if (tjDecompress2(decompressHandle, jpeg_data, jpeg_size, rgbaBuf, width, 0, height, TJPF_RGBA, TJFLAG_ACCURATEDCT) != 0)
  {
    free(rgbaBuf);
    tjDestroy(decompressHandle);
    return NULL;
  }

  tjDestroy(decompressHandle);

  uint8_t *rotated = rotate_rgba_image(rgbaBuf, width, height, rotationDegrees);
  free(rgbaBuf);

  if (!rotated)
    return NULL;

  int outWidth = (rotationDegrees == 90 || rotationDegrees == 270) ? height : width;
  int outHeight = (rotationDegrees == 90 || rotationDegrees == 270) ? width : height;

  uint8_t *jpeg = encode_rgba_to_jpeg_buffer(rotated, outWidth, outHeight, 90, out_size);
  free(rotated);
  return jpeg;
}
